### ISSUE TITLE
AspectJ based load time weaving using an LTW aware classloader without Java agent on "java -jar app.jar"

### ISSUE THREAD

**Issue by alex-konkov** [spring-projects/spring-boot#739]
AspectJ based load time weaving (LTW) can be set up with Spring Boot, however when running as a dead simple "java -jar app.jar" the classes are not woven until Java agent is specified. (AspectJ + LTW + self contained JAR) is a potentially cool combination for many Spring Boot powered projects that need to leverage AOP without needing to play any additional tricks (or worrying about not forgetting to).

This feature request is about adding an AspectJ LTW bootstrapping code straight into the Spring Boot launcher where a specific class-loader is initialised. Then, the client projects that would like to stick with the simplest "java -jar app.jar" startup way would be able to leverage power of AspectJ LTW weaving feature out of the box.

http://stackoverflow.com/questions/23219772/spring-boot-and-aspectj-based-ltw-without-java-agent/23248830?noredirect=1#comment35709788_23248830

**Comment by beihaifeiwu** [comment]
+1 to this

**Comment by kjozsa** [comment]
I would also need this feature, count me in please. +1

**Comment by kjozsa** [comment]
Also, a quick comment above as "being consistent" - if you have a spring-boot project with Eclipselink which does require LTW, running it with "mvn spring-boot:run" will work and running the packaged jar file with "java -jar" will just fail. 

An example of this problem is demonstrated with this project: https://github.com/dsyer/spring-boot-sample-data-eclipselink - clone it, mvn package and java -jar the built jar, it fails badly.

**Comment by wilkinsona** [comment]
It only works with `mvn spring-boot:run` because the [agent is configured in the `pom.xml`](https://github.com/dsyer/spring-boot-sample-data-eclipselink/blob/master/pom.xml#L83-L85):

```
<configuration>
    <agent>${settings.localRepository}/org/springframework/spring-instrument/${spring.version}/spring-instrument-${spring.version}.jar</agent>
</configuration>
```

If you don't supply the equivalent configuration when running the application using `java -jar` it will, quite reasonably, fail.

**Comment by imod** [comment]
is there anything we can do to make this happen?

**Comment by wilkinsona** [comment]
I've yet to see a compelling argument against my [usability concerns described above](https://github.com/spring-projects/spring-boot/issues/739#issuecomment-58511066).

**Comment by imod** [comment]
hmm, not sure the miss of a feature on the test framework is a compelling argument...

For us, this missing feature means we can't use load time weaving on cloudfoundry with any spring boot application. Because the deploy unit is a single JAR - if we wanna change this, then we have to fork the java_buildpack and add this functionality (that's nothing I really wanna do). 
Looking at the importance of spring for cloudfoundry (in both ways) I would assume this argument is stronger then the one about the test framework.

**Comment by wilkinsona** [comment]
> hmm, not sure the miss of a feature on the test framework is a compelling argument...

As I said above, that's a secondary concern. My primary concern is about the complexity of having things behave differently when you run using `java -jar`, vs launching the main method in your IDE, vs launching via our Maven or Gradle plugins. Rather than having to document different behaviour depending on how you launch your application, I like the consistency of the current behaviour. As things stand, you consistently have to configure the agent. Adding some magic to `java -jar` would break that consistency.

The Java buildpack already has support for a number of Java agents. Another way to tackle your particular problem would be via an enhancement to the buildpack. /cc @cgfrost

**Comment by lukiano** [comment]
This is what I did to make it work:

*pom.xml*

```xml
<plugin>
  <groupId>org.springframework.boot</groupId>
  <artifactId>spring-boot-maven-plugin</artifactId>
  <configuration>
    <layout>ZIP</layout>
    <mainClass>org.springframework.boot.SpringApplication</mainClass>
  </configuration>
</plugin>
```
ZIP layout so that it uses the `org.springframework.boot.loader.PropertiesLauncher`

*application.properties*
```properties
loader.classLoader=packageName.TransformingClassLoader
loader.main=org.springframework.boot.SpringApplication
spring.main.sources=packageName.Config
```
Here I tell the PropertiesLauncher to use my own classLoader, and that the main program is the standard `SpringApplication`, with `packageName.Config` as my Spring configuration file.

*TransformingClassLoader.java*
https://gist.github.com/lukiano/114cb01a59ddd9880aa7
based on `org.springframework.instrument.classloading.ShadowingClassLoader` and `org.springframework.core.OverridingClassLoader`

And then my Config file implements LoadTimeWeavingConfigurer
```java
public LoadTimeWeaver getLoadTimeWeaver() {
  return new ReflectiveLoadTimeWeaver(Thread.currentThread().getContextClassLoader());
}
```

So basically it loads the entire Spring context under my own ClassLoader which implements the transformer methods that `ReflectiveLoadTimeWeaver` expects. And it still works when it creates a fat-jar, so it doesn't depend on being run with `mvn spring-boot:run`

**Comment by imod** [comment]
@lukiano nice, that looks like a solution which would also solve the concerns @wilkinsona has, with some further work this might be doable even without the ZIP packaging and therefore integrate it into boot directly.
I just found an other solution which might be interesting too: https://github.com/subes/invesdwin-instrument

**Comment by lukiano** [comment]
I see, that one uses tools.jar to inject the agent itself by retrieving the jvm's process id. A solution I considered before, but discarded to avoid a dependency on the JDK

**Comment by laxika** [comment]
+1 it's really needed if possible

**Comment by jasperblues** [comment]
+1

**Comment by heroicefforts** [comment]
+1

**Comment by juzerali** [comment]
+1 count me in. I want to use Spring's [Cache Abstraction](https://docs.spring.io/spring/docs/current/spring-framework-reference/html/cache.html) in aspectj mode inside a Spring boot web application with embedded tomcat. But because I can't use aspectj load time weaving, I can't use the `@Cacheable` annotation if I am to call the method from within the service class. This sets me up for a round of refactoring to ensure that  a `@Cacheable` method is always called from a proxy.

**Comment by manojtailor** [comment]
@lukiano  : i tried the same thing which you tried to make it work but I am still unable to make aspectJ working.

Getting following error;
org.springframework.context.ApplicationContextException: Unable to start embedded container; nested exception is org.springframework.context.ApplicationContextException: Unable to start EmbeddedWebApplicationContext due to missing EmbeddedServletContainerFactory bean.

**Comment by lukiano** [comment]
@manojtailor that doesn't look like an AspectJ issue

**Comment by wilkinsona** [comment]
@manojtailor That looks like you're missing a dependency on Tomcat, Undertow, or Jetty. `spring-boot-starter-web` is the typical way to add such a dependency.

**Comment by arikkfir** [comment]
+1 from us too - would be great if we could avoid the proxy-based AOP that Spring uses by default.

**Comment by candrews** [comment]
I think a really nice solution that should make this scenario "just work" would be to make the various Spring classloaders implement org.springframework.instrument.classloading.LoadTimeWeaver. With that change, org.springframework.instrument.classloading.ReflectiveLoadTimeWeaver would work making @EnableLoadTimeWeaving work as expected (and in turn make not only AspectJ weaving work, but any other load time weaving as well).

I think these classloaders need to be modified:
* https://github.com/spring-projects/spring-boot/blob/v1.4.0.M3/spring-boot-devtools/src/main/java/org/springframework/boot/devtools/restart/classloader/RestartClassLoader.java
* https://github.com/spring-projects/spring-boot/blob/v1.4.0.M3/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/LaunchedURLClassLoader.java

**Comment by wilkinsona** [comment]
@candrews Unfortunately,  that wouldn't help when you're running the app in your IDE without DevTools as neither `RestartClassLoader` not `LaunchedURLClassLoader` is involved.

**Comment by candrews** [comment]
@wilkinsona for such cases, how about using the a simple LoadTimeWeaver classloader when not using the Restarter in SpringApplication?

**Comment by wilkinsona** [comment]
It's an interesting suggestion, but I think it would require `SpringApplication` to know about parts of the system that it shouldn't know about. For example, it would need to understand that the Launcher may or may not be used and that `DevTools` may or may not be used.

There may also be a problem with making the decision in `SpringApplication`. Once it's been called, some of the application's classes have already been loaded. Loading the same classes on two separate ClassLoaders is likely to be painful.

**Comment by tjormola** [comment]
+1

**Comment by crosslover** [comment]
+1, really need...

**Comment by Akarshit** [comment]
+1

**Comment by costap** [comment]
+1

**Comment by IvyLea** [comment]
+1

**Comment by Ant0n007** [comment]
+1

**Comment by Adobee** [comment]
+1

**Comment by philwebb** [comment]
No more `+1` comments please. They just make the issue harder to read. Just click the "thumbs up" reaction on the top comment if you want to vote.

**Comment by mauromol** [comment]
> As I said above, that's a secondary concern. My primary concern is about the complexity of having things behave differently when you run using `java -jar`, vs launching the main method in your IDE, vs launching via our Maven or Gradle plugins. 

But what about applying also the following:
- enhancing the Gradle org.springframework.boot Plugin to add the necessary references to the agent
- decorating the Spring Boot App Run Configuration in Eclipse to add the necessary references to the agent
In this way, the behaviour would be aligned between "java -jar", Gradle and Eclipse launching of the application.
Perhaps something similar may be achieved with Maven and other IDEs...

**Comment by ade90036** [comment]
+1

**Comment by vangogh-ken** [comment]
Now have to use javaagent instead

**Comment by ade90036** [comment]
It was a very painful process in getting this working before and enabling
javaagent was the only possible option (not my favourite).

Not sure how relevant in getting load time AOP working anymore.

Considering all the progress that spring has achieved in switching to AOT
compilation for spring boot 2.6 and 3.0 and hence native-image support
Maybe now aspects can be easily compiled ahead of time and easily included
in the application.

**Comment by ade90036** [comment]
Let me add some context.

I'm only using AspectJ with spring boot to enable ***@***.*** @Evict
@Scheduled @Asych` annotation to get around at a limitation of spring proxy
where in certain case a user can bypass the triggering of the annotation
and calling the plain method.

Sorry I didn't mean to inconvenient this conversation with technologies *we*
fully fully grasp. But unfortunately that is the directions that everything
is moving to (compile time, spring boot, micronaut, quarkus, native-image
Graalvm)

Which makes you wonder if there is place for AspectJ introspection if we,
for example, can create physical classes at compile time to pretty much
model the intended behaviour without needing reflection or introspection
(like a spring proxy) and added overheads... (Significant)

micronaut and quarkus are already proving this, spring boot is catching up.

Going back to my problem:

- Adding `javaagent` was the easiest thing I have done. But that also open
a can of worms for security reason, I won't get in details.

-  jar specified in javaagent needed to be available before the application
launched.

- Hence I couldn't use spring boot gradle plugin to build a fat jar

- Hence it had to be maintain a static jar stored in my project

- Therefore I lost the ability to use use gradle to track version
compatibility with my current version of spring and adjacent dependencies

- because it was a static external dependency I couldn't use convenient
spring boot / gradle plugins to generate docker image out of the box.

- I had to build a custom docker image to add static files

Now, i hope I'm not the only one to realise that all these "extra" steps,
for their intended purpose, are only in place to enable aspectJ working and
are another thing that can break down..










Hence is couldn't use any of the cradle plugin to build a dockerised image

Hence

**Comment by kriegaex** [comment]
Hi all!

If you are running your applications as **executable JARs on JRE 9+** (Spring Boot ones or other types), probably [Agent Embedder Maven Plugin](https://github.com/dev-aspectj/agent-embedder-maven-plugin?tab=readme-ov-file) is what you want.

It enables you to embed a java agent in your executable JAR and have it started automatically using the [`Launcher-Agent-Class` JVM mechanism](https://docs.oracle.com/javase/9/docs/api/java/lang/instrument/package-summary.html).

Unique features added by this plugin and unavailable via the original JVM mechanism: You can
  * embed and run **multiple agents** (the JVM supports only one out of the box),
  * pass an **option string** to each agent, just like from the JVM command line.

Spoiler: I am the author of the plugin and also happen to be the current maintainer of AspectJ.
